### PR TITLE
(Dated) Add Generated contract cache.

### PR DESCRIPTION
**Pull request by olegnat**
Added DB based contacts cache.

### CODE REVIEW

File: src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java [hunk 1]
```diff
@@ -53,27 +56,41 @@
...
     this.contractDbService = contractDbService;
     this.abiProviderService = new AbiProviderService();
+    this.contractSourceCodeRepository = contractSourceCodeRepository;
+    this.appProperties = appProperties;
   }
 
   public GeneratedContract getContract(String address, Long block, String selector,
       String network) {
+
+    Boolean isUseCache = appProperties.isUseContractSourceCache();
```
Comment (belbix) [inline src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java:83]: I guess we can use it without adding as attribute for each method

File: src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java [hunk 2]
```diff
@@ -103,26 +120,42 @@ private boolean equalContracts(GeneratedContract newContract, GeneratedContract
...
     boolean etherscanIsProxy = false;
     String etherscanProxyImpl = "";
-
-    AbiProviderService.SourceCodeResult sourceCode =
-        abiProviderService.contractSourceCode(
+    AbiProviderService.SourceCodeResult sourceCode;
+    if (isUseCache) {
+      ContractSourceCodeDTO cashed_source =
+          contractSourceCodeRepository.findByAddressNetwork(address, network);
+      if (cashed_source != null) {
```
Comment (belbix) [inline src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java:136]: please, use camelCase for variables

File: src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java [hunk 3]
```diff
@@ -103,26 +120,42 @@ private boolean equalContracts(GeneratedContract newContract, GeneratedContract
...
+      } else {
+        sourceCode = abiProviderService.contractSourceCode(
             address, networkProperties.get(network).getAbiProviderKey(), network);
+        ContractSourceCodeDTO csdto = ContractSourceModelConverter.toDTO(sourceCode);
+        csdto.setAddress(address);
+        csdto.setNetwork(network);
+        contractSourceCodeRepository.save(csdto);
+      }
+    } else {
+      sourceCode = abiProviderService.contractSourceCode(
```
Comment (belbix) [inline src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java:148]: duplicate code, we can use simplier if statement

File: src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java [hunk 4]
```diff
@@ -103,26 +120,42 @@ private boolean equalContracts(GeneratedContract newContract, GeneratedContract
...
-        abiProviderService.contractSourceCode(
+    AbiProviderService.SourceCodeResult sourceCode;
+    if (isUseCache) {
+      ContractSourceCodeDTO cashed_source =
+          contractSourceCodeRepository.findByAddressNetwork(address, network);
+      if (cashed_source != null) {
+        log.info("Used cached contract sources code {}", cashed_source.getContractName());
+        sourceCode = ContractSourceModelConverter.toSourceCodeResult(cashed_source);
+      } else {
+        sourceCode = abiProviderService.contractSourceCode(
```
Comment (belbix) [inline src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java:140]: sourceCode can be null

File: src/test/java/pro/belbix/ethparser/codegen/SimpleContractGeneratorTest.java [hunk 5]
```diff
@@ -90,4 +88,17 @@ void getContract_ZeroEx() {
         () -> assertEquals(5, contract.getFunctions().size(), "Functions size")
     );
   }
+
```
Comment (belbix) [inline src/test/java/pro/belbix/ethparser/codegen/SimpleContractGeneratorTest.java:91]: please add a test for contract with empty source code
example https://etherscan.io/address/0x0000000000007f150bd6f54c40a34d7c3d5e9f56#code

File: src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java (hunk 1 above)
Comment (olegnat) [reply src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java:83]: ACK

File: src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java (hunk 2 above)
Comment (olegnat) [reply src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java:136]: Python legacy. ACK

File: src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java (hunk 3 above)
Comment (olegnat) [reply src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java:148]: Ack

File: src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java (hunk 4 above)
Comment (olegnat) [reply src/main/java/pro/belbix/ethparser/codegen/SimpleContractGenerator.java:140]: Ack

File: src/test/java/pro/belbix/ethparser/codegen/SimpleContractGeneratorTest.java (hunk 5 above)
Comment (olegnat) [reply src/test/java/pro/belbix/ethparser/codegen/SimpleContractGeneratorTest.java:91]: ack
